src/test/objectstore/store_test.cc

```diff
@@ -8638,6 +8638,152 @@ TEST_P(StoreTestSpecificAUSize, DeferredDifferentChunks) {
   }
 }
 
+TEST_P(StoreTestSpecificAUSize, DeferredAndClone) {
+
+  if (string(GetParam()) != "bluestore")
+    return;
+
+  size_t alloc_size = 4096;
+  size_t prefer_deferred_size = 65536;
+ 
+  SetVal(g_conf(), "bluestore_block_db_create", "true");
+  SetVal(g_conf(), "bluestore_block_db_size", stringify(1 << 30).c_str());
+  
+  StartDeferred(alloc_size);
+  SetVal(g_conf(), "bluestore_prefer_deferred_size",
+    stringify(prefer_deferred_size).c_str());
+  g_conf().apply_changes(nullptr);
+
+  int r;
+  coll_t cid;
+
+  ghobject_t hoid(hobject_t("test", "", CEPH_NOSNAP, 0, -1, ""));
+  hoid.hobj.pool = -1;
+  ghobject_t hoid2(hobject_t(sobject_t("Object 2", CEPH_NOSNAP)));
+  hoid2.hobj.pool = -1;
+
+  ObjectStore::CollectionHandle ch = store->create_new_collection(cid);
+  {
+    ObjectStore::Transaction t;
+    t.create_collection(cid, 0);
+    t.touch(cid, hoid);
+    r = queue_transaction(store, ch, std::move(t));
+    ASSERT_EQ(r, 0);
+  }
+  {
+    ObjectStore::Transaction t;
+    bufferlist bl;
+    bl.append(std::string(3, 'z'));
+    t.write(cid, hoid, 0, bl.length(), bl,
+            CEPH_OSD_OP_FLAG_FADVISE_NOCACHE);
+    r = queue_transaction(store, ch, std::move(t));
+    ASSERT_EQ(r, 0);
+  }
+  {
+    ObjectStore::Transaction t;
+    t.clone_range(cid, hoid, hoid2, 0, 3, 0);
+    cerr << "Clone range object" << std::endl;
+    r = queue_transaction(store, ch, std::move(t));
+    ASSERT_EQ(r, 0);
+ }
+ sleep(1);
```

Comment:
What are the chances of the deferred write finishing before `clone_range` transaction takes place? Since I expected this to be undeterministic I relied on Synthetic more. If this still remains undeterministic it might be kinda useful to do this in a tight loop to maximize chances of this ocurring.

Reply:
Well, deferred write completion (at least in this specific case when there is no other load) is determined by bluestore_max_defer_interval (=3 sec by default).

Which means that clone_range() has pretty good chances to complete first. And in this case we'll get the behavior we expect for this test case.
In an unlikely case If that doesn't happen the test case would silently succeed without running through the expected op sequence - sort of false positive behavior. Which is not too bad if it happens rarely. So 
The rationale why I don't like to have some loop here or even rely on Syntetic is that they are bulky and hard to debug if the problem pops up.

sleep(1) call here was intended to make sure (a bit unreliably) that following read happens after clone_range. Which is crucial as failing to do that would fail the UT. I missed that I can do with ch.wait() call though - will update shortly.